### PR TITLE
add docs on how to define kaniko volume mount and update secret instructions

### PR DESCRIPTION
**Pull request by tejal29**
Fixes: #4175

Add docs on 
1. How to configure volume mounts for kaniko pod
2. Better secret docs.

### CODE REVIEW

File: docs/content/en/docs/pipeline-stages/builders/docker.md [hunk 1]
```diff
@@ -67,15 +67,37 @@ These credentials are configured in the `cluster` section with the following opt
...
+in the Google Cloud Console project you want to push the final image to with Storage Admin permissions. 
+Download a JSON key for this service account at a convenient location.
+```yaml
+build:
+  cluster:
+    pullSecretPath: path-to-local-service-account-key-file
 ```
+Skaffold will create a secret from the service account key file. Skaffold will delete the secret from your cluster at the end of the build.
+
+(**Note**: Please do not check this service account key file in your git repository)
```
Comment (briandealwis) [inline docs/content/en/docs/pipeline-stages/builders/docker.md:98]: The dev style guide [advocates against _please_](https://developers.google.com/style/tone#politeness-and-use-of-please).  Besides, this seems worthy of being an order :-)

```suggestion
(**Note**: Do not check this service account key file in your git repository)
```

File: docs/content/en/docs/pipeline-stages/builders/docker.md [hunk 2]
```diff
@@ -67,15 +67,37 @@ These credentials are configured in the `cluster` section with the following opt
...
+(**Note**: Rename the downloaded JSON key to *kaniko-secret* without appending *.json*).
+
+Alternatively, the path to credentials file can be set with the `pullSecretPath` option:
 ```yaml
 build:
   cluster:
     pullSecretName: pull-secret-in-kubernetes
-    # OR
-    pullSecret: path-to-service-account-key-file
+    pullSecretPath: path-to-service-account-key-file-within-secret
```
Comment (briandealwis) [inline docs/content/en/docs/pipeline-stages/builders/docker.md:81]: `within`?  `with` might be better?

File: docs/content/en/docs/pipeline-stages/builders/docker.md [hunk 3]
```diff
@@ -67,15 +67,37 @@ These credentials are configured in the `cluster` section with the following opt
 
 {{< schema root="ClusterDetails" >}}
 
-To set up the credentials for Kaniko refer to the [kaniko docs](https://github.com/GoogleContainerTools/kaniko#kubernetes-secret) (**Note**: Rename the downloaded JSON key to *kaniko-secret* without appending *.json*).
-Alternatively, the path to a credentials file can be set with the `pullSecret` option:
+### Configure Kaniko Credentials 
+
+To set up the credentials for Kaniko refer to the [kaniko docs](https://github.com/GoogleContainerTools/kaniko#kubernetes-secret)
+
+(**Note**: Rename the downloaded JSON key to *kaniko-secret* without appending *.json*).
```
Comment (briandealwis) [inline docs/content/en/docs/pipeline-stages/builders/docker.md:74]: Should this be placed in the artifact's `context` directory?

File: docs/content/en/docs/pipeline-stages/builders/docker.md [hunk 4]
```diff
@@ -67,15 +67,37 @@ These credentials are configured in the `cluster` section with the following opt
...
   cluster:
     pullSecretName: pull-secret-in-kubernetes
-    # OR
-    pullSecret: path-to-service-account-key-file
+    pullSecretPath: path-to-service-account-key-file-within-secret
+  
+```
+
+Skaffold can also set up credentials if a secret does not exist in your cluster. This is usually the case when you are performing a 
+build in your personal cluster.
```
Comment (briandealwis) [inline docs/content/en/docs/pipeline-stages/builders/docker.md:86]: To me, `personal cluster` suggests Minikube or something similar.

File: docs/content/en/docs/pipeline-stages/builders/docker.md [hunk 5]
```diff
@@ -67,15 +67,37 @@ These credentials are configured in the `cluster` section with the following opt
...
+Skaffold can also set up credentials if a secret does not exist in your cluster. This is usually the case when you are performing a 
+build in your personal cluster. 
+
+First, [create a service account](https://cloud.google.com/iam/docs/creating-managing-service-accounts#creating) 
+in the Google Cloud Console project you want to push the final image to with Storage Admin permissions. 
+Download a JSON key for this service account at a convenient location.
+```yaml
+build:
+  cluster:
+    pullSecretPath: path-to-local-service-account-key-file
```
Comment (briandealwis) [inline docs/content/en/docs/pipeline-stages/builders/docker.md:94]: It's unfortunate this is called `pull`?

File: docs/content/en/docs/pipeline-stages/builders/docker.md (hunk 3 above)
Comment (tejal29) [reply docs/content/en/docs/pipeline-stages/builders/docker.md:74]: no. i added a note down that please do not check in secret in the directory. This secret could be at any location.

File: docs/content/en/docs/pipeline-stages/builders/docker.md (hunk 2 above)
Comment (tejal29) [reply docs/content/en/docs/pipeline-stages/builders/docker.md:81]: actually it is within the secret. 
e.g.
```
kubectl create secret generic kaniko-secret --from-file=~/kaniko-secret.json`
```
The secret created will be loaded at path `kaniko-secret.json`.

File: docs/content/en/docs/pipeline-stages/builders/docker.md (hunk 4 above)
Comment (tejal29) [reply docs/content/en/docs/pipeline-stages/builders/docker.md:86]: it could be your personal (local or remote) GKE cluster  In case developers are using. 
Is there an umbrella term for that?

File: docs/content/en/docs/pipeline-stages/builders/docker.md (hunk 5 above)
Comment (tejal29) [reply docs/content/en/docs/pipeline-stages/builders/docker.md:94]: This secret is only used for pulling images and we construct `GoogleApplicationCredentials` from it. 
Hence it was named as pull secret originally. But yes something we could re-think

File: docs/content/en/docs/pipeline-stages/builders/docker.md (hunk 3 above)
Comment (briandealwis) [reply docs/content/en/docs/pipeline-stages/builders/docker.md:74]: Sorry, I'm not very adept remote cluster builds with Kaniko, and I'm confused by both this section and the linked to Kaniko secrets.  The linked-to Kaniko document uses kubectl to create a secret wrapping the service-account key file to access GCR.

Would it make sense to just copy the relevant steps here?

It's also not clear to me where I'd put this `kaniko-secret` file.  It seems there's a standard place, given the "Alternatively, …" sentence below.

File: docs/content/en/docs/pipeline-stages/builders/docker.md [hunk 6]
```diff
@@ -67,15 +67,37 @@ These credentials are configured in the `cluster` section with the following opt
...
 
-To set up the credentials for Kaniko refer to the [kaniko docs](https://github.com/GoogleContainerTools/kaniko#kubernetes-secret) (**Note**: Rename the downloaded JSON key to *kaniko-secret* without appending *.json*).
-Alternatively, the path to a credentials file can be set with the `pullSecret` option:
+### Configure Kaniko Credentials 
+
+To set up the credentials for Kaniko refer to the [kaniko docs](https://github.com/GoogleContainerTools/kaniko#kubernetes-secret)
+
+(**Note**: Rename the downloaded JSON key to *kaniko-secret* without appending *.json*).
+
+Alternatively, the path to credentials file can be set with the `pullSecretPath` option:
```
Comment (briandealwis) [inline docs/content/en/docs/pipeline-stages/builders/docker.md:76]: What is this alternative to?  The `(**Note**: …` or the `To set up the credentials …`?

File: docs/content/en/docs/pipeline-stages/builders/docker.md (hunk 4 above)
Comment (briandealwis) [reply docs/content/en/docs/pipeline-stages/builders/docker.md:86]: I think it would be helpful to explicitly highlight that this is for GKE/GCR?

File: docs/content/en/docs/pipeline-stages/builders/docker.md (hunk 3 above)
Comment (chanseokoh) [reply docs/content/en/docs/pipeline-stages/builders/docker.md:74]: No wonder it's confusing.

When creating a secret with
```
kubectl create secret generic kaniko-secret --from-file=<path to kaniko-secret.json>
```
the key name of the `data:` map in the created secret is set to the filename given by `--from-file`. So, if you follow the linked Kaniko doc, you'll end up having the following for the secret.
```yaml
data:
  kaniko-secret.json: <... secret data ...>
``` 
However, Skaffold by default [expects](https://github.com/GoogleContainerTools/skaffold/issues/3828#issuecomment-626965077) you to create the secret with the key name `kaniko-secret`. That's the reason for the instruction to rename the JSON key file to `kaniko-secret` without `.json`.

File: docs/content/en/docs/pipeline-stages/builders/docker.md (hunk 6 above)
Comment (chanseokoh) [reply docs/content/en/docs/pipeline-stages/builders/docker.md:76]: The former (i.e., `**Note**`).

Recently `pullSecretPath` was converted to serve as a dual-purpose property (#4381) to address the issue that Skaffold uses the hard-coded secret key name `kaniko-secret`. So, instead of renaming the JSON key file to `kaniko-secret` (without `.json`), you can now say, e.g., `pullSecretPath: my-key.json` if you created your secret with
```
kubectl create secret generic kaniko-secret --from-file=.../my-key.json
```
Another alternative not documented here is to explicitly specify the key name with
```
kubectl create secret generic kaniko-secret --from-file=kaniko-secret=.../my-key.json
```

File: docs/content/en/docs/pipeline-stages/builders/docker.md (hunk 3 above)
Comment (tejal29) [reply docs/content/en/docs/pipeline-stages/builders/docker.md:74]: > It's also not clear to me where I'd put this `kaniko-secret` file. It seems there's a standard place, given the "Alternatively, …" sentence below.

We could copy the instructions here instead of pointing to kaniko docs.

File: docs/content/en/docs/pipeline-stages/builders/docker.md (hunk 4 above)
Comment (tejal29) [reply docs/content/en/docs/pipeline-stages/builders/docker.md:86]: Kaniko can also on EKS, Azure hence i mentioned personal cluster. 
What i want to focus on is, usually if you are in a shared cluster env, secrets will be installed by admins and developers don't have access to create them

However, when you are using a personal cluster to deploy or in personal namespace on a shared cluster, you would need to create you own secret.

File: docs/content/en/docs/pipeline-stages/builders/docker.md (hunk 3 above)
Comment (MarlonGamez) [reply docs/content/en/docs/pipeline-stages/builders/docker.md:74]: I agree that we should copy the instructions to these docs. It makes for a much easier read through them

File: docs/content/en/docs/pipeline-stages/builders/docker.md (hunk 4 above)
Comment (gsquared94) [reply docs/content/en/docs/pipeline-stages/builders/docker.md:86]: maybe reword to :
```
Skaffold can also set up credentials if a secret does not exist in your cluster. This is usually the case when you are not using a shared cluster where the cluster admin would have already configured it for you. 
```

File: docs/content/en/docs/pipeline-stages/builders/docker.md [hunk 7]
```diff
@@ -94,6 +116,33 @@ Docker image `gcr.io/k8s-skaffold/example` with Kaniko:
...
+      cache:
+        repo: getting-started
+      volumeMounts:
+      - name: kaniko-cache
+        mountpath: /cache
+        readonly: true
+cluster:
+  volumes:
+  - name: kaniko-cache
+    persistentvolumevlaim:
```
Comment (briandealwis) [inline docs/content/en/docs/pipeline-stages/builders/docker.md:140]:
```suggestion
    persistentvolumeclaim:
```

Comment (briandealwis) [reply docs/content/en/docs/pipeline-stages/builders/docker.md:140]: I think we need to retrieve this from a file so that we can do some validation as I think this example need another level of indirection for `volumesource` and should actually be:

```
    volumes:
      - name: kaniko-cache
        volumesource:
          persistentvolumeclaim: 
            claimname: kaniko-cache-claim
```
